File: eulerglot/javascript/src/001Multiples3And5.js
function buildArray(arr, floor, ceil) {
  arr.push(floor)
  if (floor == ceil) {
    return arr;
  }
  return buildArray(arr, floor + 1, ceil);
}

function sumUnder(ceil) {
  array = buildArray([], 1, ceil - 1);
  return sum(array, 0)
}

function sum(array, acc) {
  if (array.length == 0) {
    return acc;
  }

  number = array[array.length - 1]

  if (number % 3 == 0 || number % 5 == 0) {
    return sum(tail(array), number + acc);
  }

  return sum(tail(array), acc);
}

function tail(array) {
  return array.slice(0, array.length - 1);
}

module.exports = {
  buildArray: buildArray,
  sumUnder: sumUnder,
  sum: sum,
  tail: tail
}

File: eulerglot/clojure/src/clojure_euler/001_multiples_3_and_5.clj
(ns clojure-euler.001_multiples_3_and_5)

(comment
  "If we list all the natural numbers below 10 that are multiples of 3 or 5,
  we get 3, 5, 6 and 9. The sum of these multiples is 23.
  Find the sum of all the multiples of 3 or 5 below 1000.")

(defn count-multiples-of
  [multiple number]
  (int (Math/floor (/ number multiple))))

(defn sum-multiples-of
  [multiple number]
  (let [countm (count-multiples-of multiple number)]
  (let [calc   (/ (+ multiple (* countm multiple)) 2.0)]
  (* countm calc))))

(defn sum_under
  [ceil]
  (let [number (- ceil 1)]
  (let [sum3   (sum-multiples-of 3 number)]
  (let [sum5   (sum-multiples-of 5 number)]
  (let [sum15  (sum-multiples-of 15 number)]
  (int (- (+ sum3 sum5) sum15)))))))
